File: app/src/main/java/com/sevtinge/hyperceiler/module/hook/securitycenter/battery/UnlockSmartCharge.java
package com.sevtinge.hyperceiler.module.hook.securitycenter.battery;

import static com.sevtinge.hyperceiler.module.base.tool.HookTool.MethodHook.returnConstant;

import android.util.SparseArray;

import com.sevtinge.hyperceiler.module.base.BaseHook;
import com.sevtinge.hyperceiler.module.base.dexkit.DexKit;
import com.sevtinge.hyperceiler.module.base.dexkit.IDexKit;
import com.sevtinge.hyperceiler.module.base.dexkit.IDexKitList;

import org.luckypray.dexkit.DexKitBridge;
import org.luckypray.dexkit.query.FindMethod;
import org.luckypray.dexkit.query.matchers.AnnotationMatcher;
import org.luckypray.dexkit.query.matchers.AnnotationsMatcher;
import org.luckypray.dexkit.query.matchers.MethodMatcher;
import org.luckypray.dexkit.result.MethodData;
import org.luckypray.dexkit.result.MethodDataList;

import java.lang.reflect.AnnotatedElement;
import java.lang.reflect.Method;
import java.util.List;

import de.robv.android.xposed.XC_MethodHook;

public class UnlockSmartCharge extends BaseHook {
    @Override
    public void init() throws NoSuchMethodException {
        List<Method> methods = DexKit.getDexKitBridgeList("VendorSmartChg", new IDexKitList() {
            @Override
            public List<AnnotatedElement> dexkit(DexKitBridge bridge) throws ReflectiveOperationException {
                MethodDataList methodData = bridge.findMethod(FindMethod.create()
                        .matcher(MethodMatcher.create()
                                .usingStrings("persist.vendor.smartchg")
                        )
                );
                return DexKit.toElementList(methodData, lpparam.classLoader);
            }
        }).toMethodList();
        for (Method method : methods) {
            hookMethod(method, returnConstant(true));
        }
    }
}

File: app/src/main/java/com/sevtinge/hyperceiler/module/hook/securitycenter/battery/UnlockSmartCharge.kt
package com.sevtinge.hyperceiler.module.hook.securitycenter.battery

import com.github.kyuubiran.ezxhelper.ClassLoaderProvider.classLoader
import com.github.kyuubiran.ezxhelper.HookFactory.`-Static`.createHooks
import com.sevtinge.hyperceiler.module.base.BaseHook
import com.sevtinge.hyperceiler.utils.DexKit.addUsingStringsEquals
import com.sevtinge.hyperceiler.utils.DexKit.dexKitBridge

object UnlockSmartCharge : BaseHook() {
    private val smartChg by lazy {
        dexKitBridge.findMethod {
            matcher {
                addUsingStringsEquals("persist.vendor.smartchg")
            }
        }.map { it.getMethodInstance(classLoader) }.toList()
    }

    override fun init() {
        smartChg.createHooks {
            returnConstant(true)
        }
    }
}
